### ISSUE TITLE
XMLHTTPRequest set Cookie and read Set-Cookie

### ISSUE THREAD

**Issue by philcali** [pebble/pebblejs#76]
I'm not sure if this is a pebblejs issue, or a Pebble JS runtime issue on the phone; I've had little luck on the SDK forum, so I'm x-posting here:

I'm aware of the forbidden headers in the XHR spec: https://fetch.spec.whatwg.org/#forbidden-header-name

There is two distinct problems with this, however:

1.  In development, the emulator CAN set Cookie's and read Set-Cookie's. I imagine this is because the underlying implementation of XMLHTTPRequest in the emulator is python's urllib or something similar.
2. JS runtimes on the phone CANNOT set Cookie's and read Set-Cookie's using the same code.

There is a disconnect between the two, but I believe that the phone's JS runtime should support general HTTP interaction as if your were interacting with some HTTPConnection. Is there a simple configuration flag I can put on the XMLHTTPRequest object call to support these headers? I haven't had much success with just `withCredentials = true`, either.

Thanks for your help!

**Comment by philcali** [comment]
I imagine the use case outlined is so rare that it's simply not worth the time and energy to investigate. Since I've found a decent enough workaround for my problem, I'm closing this issue.

For those interested in the workaround: don't rely on cookies! If you are integrating with a third party service that relies on cookies (like I was), your best bet is to build a service to proxy the calls with the authenticated session for you. On the plus side, it makes web requests initiated by Pebble very simple, as it was intended to be.

**Comment by pafcu** [comment]
I ran into the same problem as you, but I disagree about the solution to use a proxy: I really don't want to handle other people's passwords. I'd rather just send them directly to the correct server (not controlled by me). It also means I have to run my own proxy just for a small server, and that my app stops working if I have server issues.

Maybe consider re-opening the issue?

**Comment by philcali** [comment]
I agree that it's a problem, and the workaround is not ideal. I also think the problem lies on the Pebble JS runtime on the phone, so it really falls outside of the scope of PebbleJS. Ultimately the issue falls into this technical grey area: whether or not cross origin sandboxing should stay true to the JS specification. Of course there are advantages and disadvantages in either case, but it would certainly be worth discussing to those involved in the runtime project.

As far as proxying the requests, AWS's [API Gateway](https://aws.amazon.com/api-gateway/) makes this task pretty easy. I believe once you start proxying, a world of features open up to you that were closed before.

**Comment by jwise** [comment]
+1 to reopen.  I'd really rather not handle passwords, either.

**Comment by philcali** [comment]
You can store credentials once as a Pebble configuration page and the information is stored on the watch, which is no different than having a native app that does the same thing. You can still proxy the requests with API Gateway, and accomplish this task so I don't quite understand where the "handling passwords" hold up is coming from (for you to have a session, you need the credentials anyway). That said, having a definitive answer on what _should_ happen with remote requests using a `XMLHTTPRequest` would be great, so I'm opening the ticket for that reason.

**Comment by hkpeprah** [comment]
Is the issue that you're sending a user to a third-party service to authenticate, and want to get the cookies returned in the authenticated session so that you can then send requests as if you are said user (and therefore authenticated)?  Seems like a case for something like [OAuth](https://dev.twitter.com/oauth/overview/faq), though I'm guessing this has to do with services that don't provide OAuth?

**Comment by philcali** [comment]
Correct. The original problem is trying to communicate to a third party service that doesn't support OAuth, and we don't have control over it. Ideally the service would simply support modern authentication methods and no workaround would be needed.

**Comment by jwise** [comment]
I don't mind storing credentials in the app -- in that case, they're not leaving the device.  However, I'd rather not have user credentials passing through a server that I control.

(On the other hand, in the case of the service that I am proxying for, you can call them up on the telephone, and provide the e-mail address in question and a year of birth, and they'll just read your password out to you.  But it's the principle of the matter!)

**Comment by Meiguro** [comment]
Hmm, this is a problem with the PebbleKit JS JavaScript VM as you have found, and not particular to Pebble.js, but it's relevant enough to warrant a ticket here. Have you tried contacting devsupport `at` pebble.com? That is the normal process for potential issues with the base SDK, though I am curious to this problem as well. I will write some thoughts here.

I agree that it is pretty nasty to pass credentials through your own server. If this cookie header setting restriction were removed however, I'm not sure it would solve the idea behind that nastiness -- something would need to intercept the cookie in order to pass it to your service JavaScript. And to intercept the cookie, that would mean having the potential to intercept the password.

AFAIK, the only way to guarantee that code originating from you doesn't touch, or doesn't even have the potential to touch the password is if you use the WebView directed to the service you're trying to access. But after the user logs in, something needs to grab the cookie and close the WebView. If the service is just an old fashioned cookie setting server from yesteryear's web days, ways to do so is to fake the request with your own code, or worse, encapsulate one or the other with an iframe that somehow is able to communicate which I don't think is possible with modern browsers anymore. If the user were to not trust any code that has the potential to grab their password, these methods are just as bad as a proxy from their perspective. Why they hold their password dear and not authority over the account is a different puzzle ;)

I think, then, that OAuth or some derivative that utilizes the WebView and redirects to a page after authentication is the solution, and that would be on the part of the service you're trying to access to implement unfortunately. If you do come up with a way to gain the cookie without ever having the potential to access the user password that works for all cookie based services however please do post here, though I'm not in charge of the implementations of the JavaScript VMs that are in use with PebbleKit JS.

I can think of at least one alternate method, but even then there are still pitfalls. If there was a way for a user's JavaScript to abide to a CORS-like contract -- that it can only communicate to certain blessed domains -- then on the surface it would seem to lower how much it matters that the code can intercept the password since it limits how the password can travel. But let's say the old web service is a forum. An evil JavaScript app could appear as a regular forum viewer, but can secretly post the password in a thread or over the forum's direct message functionality. Or more straightforward, the target service has a web exploit that allows outside communication. Furthermore, this also assumes that both the PebbleKit JS VM and the Pebble C are impenetrable; the C app is considered since this functionality would be available for all Pebble apps, not just Pebble.js apps. The Pebble App would also have to prompt the user each time the contract updates as well.

Without a method to obtain the cookie without having guarantees that the password cannot be transmitted or saved elsewhere, it would be ethically difficult to encourage other developers to use as a pattern in their app since it means encouraging the possibility to intercept passwords, which might be what the W3C imagined, and is of course what OAuth was made to address. With all that said however, you're free to implement a proxy to work around this issue. I can understand the frustration with being unable to access legacy systems with what seems like small restrictions, but I also see W3C's intention on the other hand.

**Comment by pafcu** [comment]
For me the problem is not the user trusting the developer. Yes, OAuth would be nicer, but we don't have it, so it's up to the user to decide it they trust the code or not.

For me the problem is the developer not trusting the developer. I'm smart enough to realize that plenty of things can go wrong when you start sending passwords around to proxy servers (whether it's a 3rd party server or something I run myself).

Using the cookie approach I need to ask the user for the password once, and after that I can throw it away, without having access to any sensitive information after that. Not because I think my code will do something bad with the login details on purpose, but because I think it might do something bad with them by accident (or due to some exploit).

Normal Javascript does NOT have access to the cookie data, so it's not even possible to intercept it. The User-agent handles all of that in the background. So after the login, from the perspective of my own code, I just need to do normal HTTP requests, and the authentication is handled behind the scenes, out of reach of my own code. The password is only needed at the time of login, and is not needed for subsequent requests.

All of this, in fact, seems to work in a normal web browser, and on iOS, but not on Android. This has been confirmed by devsupport, so they know about the issue as well.

**Comment by Meiguro** [comment]
Ah, so it works on iOS but not Android? That's interesting, I'll have to ask about how that would be resolved. When I get an answer I'll be sure to post here again.

> For me the problem is the developer not trusting the developer

In creating platforms, the worst case is taken into consideration so that they may be prevented, including misbehaving developers. That said, I surely don't think anyone posting in this ticket is a misbehaving developer.

> Not because I think my code will do something bad with the login details on purpose, but because I think it might do something bad with them by accident (or due to some exploit).

That's exactly why limited platforms exists, so these worst cases can't happen. It also results in very frustrating APIs. I was merely expanding on why the limitation exists. I'll be sure to question the discrepancy between iOS and Android since I wasn't aware of that.

> Normal Javascript does NOT have access to the cookie data, so it's not even possible to intercept it.

I was referring to `document.cookie` which only works if the page is in the same domain of course, but I admit that was silly of me to point out since the page must originate from the same domain. Sandbox breaking functionality would be needed for it, such as unrestricted iframes. However, the password being intercepted is the main concern.

I can see this feature being possibly supported because one platform allows it, but I didn't want to give false hope, so I attempted to expand on the issue. It's really a question of how far a platform should go to stop passwords from leaking. Fitbit's OAuth API even forbids the use of WebViews! It's a bit farfetched, but they're not wrong; the WebView can be hooked in many arbitrary ways. It's as good as compromised.

> [Use of WebView is prohibited. *From the Fitbit OAuth2 docs*](https://dev.fitbit.com/docs/oauth2/)

Once the problem is framed that way, you could say a platform that doesn't stop passwords from leaking is asking for it. In this particular case, the idea is whether we want to support middleman login forms. It might actually even be easier to implement than using OAuth, and could end up being used as a means when it shouldn't. Of course, this appears as a slippery slope, but I only mention this as a possibility and not an inevitability. We would have to deploy the functionality to know if that would occur at all, but the risk has to be taken into account before doing so.

In any case, it seems iOS has already taken that risk! I'll be asking around just which platform should change. It's unfortunate but the outcome could be anything, including no change. When only one platform supports a feature, developers generally avoid it since it's not fully supported. iOS is one to have very limited API's, so even if it was wanted to be disabled on iOS, it could ironically be difficult to do so. But note that I'm not in any particular position to make comment on the iOS implementation or what will get decided. I just want to give a heads up.

**Comment by pafcu** [comment]
I develop apps for my own use, so I'm not a huge fan of restrictions that are there to protect myself from me stealing my own data :-) (I am OTOH, a fan of restrictions that are there to protect myself from my own incompetence).

For my own app, I actually put a huge checkbox with a disclaimer that the user accepts being an idiot for inserting his/her password in a 3rd party app. Entering the password is only allowed if the box is checked. However, in the source code there is a switch which eliminates this step, so it's possible to bypass it by proving that you have actually looked at the source code. I'm thinking of making the source-code based switch mandatory so that simply checking the box is not enough.

So, I fully agree that users should not be typing their password in random places. It's just that sometimes that user is me, and I know exactly what the app is doing.

**Comment by jwise** [comment]
@Meiguro -- "Without a method to obtain the cookie without having guarantees that the password cannot be transmitted or saved elsewhere, it would be ethically difficult to encourage other developers to use as a pattern in their app since it means encouraging the possibility to intercept passwords, which might be what the W3C imagined, and is of course what OAuth was made to address."

This is "kind of" what the W3C imagined.  The goal in making XMLHTTPRequest cookie-unsafe is to protect web sites from other web sites ... not to protect web sites from applications.  XMLHTTPRequest is used in PebbleJS as a convenience API (exposed to applications that the user has already trusted by explicitly installing them), for compatibility with what developers already know; compare the usage of XHR on the web, where applications are expected to be untrusted, and no user input is required to execute an XHR (i.e., it could be part of a hidden iframe dropped by malware on a web page).

Similarly, the W3C is, for better or worse, generally not concerned with how web sites manage authentication credentials.  If the W3C enforced that everyone used OAuth, or secure practices in general, then indeed, we wouldn't have this problem :-) But in the mean time, we still have to live with these awful legacy APIs, and we need a way to make them work: if the remote end wanted secure remote usage, then they'd provide OAuth.

So yes, the W3C's design makes sense for the web.  When just commandeering the API for another programming interface, though, it makes sense to augment it to not strip cookies.

**Comment by Katharine** [comment]
I basically agree with jwise here (and this is an old conversation we keep come back around to — CORS enforcement was basically the thing, for instance).

This also has nothing to do with Pebble.js.

**Comment by Meiguro** [comment]
> The goal in making XMLHTTPRequest cookie-unsafe is to protect web sites from other web sites ... not to protect web sites from applications

After last night, this is what I had come to realize. Basically, if I am to view the Pebble as a computer, it should be able to have clients. And if a legacy system requires cookie based authentication, it should be possible to write a suitable client. The decision of whether a client should have access to a password should be up to the user, just like with desktop clients.

Thanks for the discussion everyone, I was viewing the issue through the lens of OAuth. That only works if everything spoke OAuth as jwise pointed out.

> This also has nothing to do with Pebble.js.

Yes, it's a bit of an issue when I get non-Pebble.js tickets. Would it make sense to have a Pebble SDK issue tracker?

**Comment by Katharine** [comment]
We're looking into it! For now, emailing devsupport gets the job done.

**Comment by robisodd** [comment]
Sorry to post here, but it seems to be the central thread for this issue.  I noticed the issue persists with v4.0.  Are there any plans to implement cookies in PebbleKit JS in the near (or not so near) future?  Just wondering if I should keep my hopes up.  I have a few projects that have fallen flat due to the lack of cookie access.

Thank you.

**Comment by laDanz** [comment]
Already another year over?
I have the same question like @robisodd:

> Are there any plans to implement cookies in PebbleKit JS in the near (or not so near) future?
